### PR TITLE
Function hoister

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libjulia/FunctionHoister.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		"{ let a:u256 function f() { let b:u256 function g() { let c:u256} let d:u256 } }",
+		"{ { let a:u256 } function g() { let c:u256 } function f() { let b:u256 let d:u256 } }"
+	);
+}
+
+BOOST_AUTO_TEST_CASE(empty_block)
+{
+	CHECK(
+		"{ let a:u256 { } function f() -> x:bool { let b:u256 := 4:u256 {} for {} f() {} {} } }",
+		"{ { let a:u256 } function f() -> x:bool {let b:u256 := 4:u256 for {} f() {} {} } }"
```
Comment (axic) [inline test/libjulia/FunctionHoister.cpp]: There should be a space between `{let` ?

File: libjulia/optimiser/FunctionHoister.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+#include <libjulia/optimiser/ASTWalker.h>
+
+namespace dev
+{
+namespace julia
+{
+
+/**
+ * Moves all functions to the top-level scope, preceded by a single block of statements.
```
Comment (chriseth) [inline libjulia/optimiser/FunctionHoister.h]: Add note: Only works on disambiguated source.

File: libjulia/optimiser/FunctionHoister.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+using namespace dev;
+using namespace dev::julia;
+using namespace dev::solidity;
+
+
+void FunctionHoister::operator()(Block& _block)
+{
+	bool topLevel = m_isTopLevel;
+	m_isTopLevel = false;
+	for (auto&& st: _block.statements)
```
Comment (axic) [inline libjulia/optimiser/FunctionHoister.cpp]: Can you use `statement` here?

File: libjulia/optimiser/FunctionHoister.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+			st = Block{_block.location, {}};
+		}
+	}
+	auto isEmptyBlock = [](Statement const& _st) -> bool {
+		return _st.type() == typeid(Block) && boost::get<Block>(_st).statements.empty();
+	};
+	if (topLevel)
+	{
+		Block firstBlock{_block.location, {}};
+		for (auto&& st: _block.statements)
```
Comment (axic) [inline libjulia/optimiser/FunctionHoister.cpp]: Same here.

File: libjulia/optimiser/FunctionHoister.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+			m_functions.emplace_back(std::move(st));
+			st = Block{_block.location, {}};
+		}
+	}
+	auto isEmptyBlock = [](Statement const& _st) -> bool {
+		return _st.type() == typeid(Block) && boost::get<Block>(_st).statements.empty();
+	};
+	if (topLevel)
+	{
+		Block firstBlock{_block.location, {}};
```
Comment (axic) [inline libjulia/optimiser/FunctionHoister.cpp]: I think this part could be a separate step. It basically wraps all root level statement which are not function definitions into a block.

Comment (chriseth) [reply libjulia/optimiser/FunctionHoister.cpp]: We can separate it, but in general, the task of the function hoister is to bring the AST into a specific normal form which is required by further steps due to C++'s memory management.

If you want, I can move that out, though. Please decide :)

Comment (axic) [reply libjulia/optimiser/FunctionHoister.cpp]: I'd prefer to be separate, it should be a tiny amount of code. Later if we profile the speed of these transformation we might decide to merge some if they perform really bad?

File: libjulia/optimiser/FunctionHoister.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		for (auto&& statement: _block.statements)
+			if (!isEmptyBlock(statement))
+				firstBlock.statements.emplace_back(std::move(statement));
+		_block.statements = vector<Statement>();
+		_block.statements.emplace_back(std::move(firstBlock));
+		std::move(m_functions.begin(), m_functions.end(), std::back_inserter(_block.statements));
```
Comment (axic) [inline libjulia/optimiser/FunctionHoister.cpp]: This should be replaced with your new `move vector`? `m_functions += _block.statements`?

File: libjulia/optimiser/FunctionHoister.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+using namespace dev;
+using namespace dev::julia;
+using namespace dev::solidity;
+
+
+void FunctionHoister::operator()(Block& _block)
+{
+	bool topLevel = m_isTopLevel;
+	m_isTopLevel = false;
+	for (auto&& statement: _block.statements)
```
Comment (pirapira) [inline libjulia/optimiser/FunctionHoister.cpp:43]: Why `&&`?  Elements of `_block.statements` have identities and are lvalues.

Comment (chriseth) [reply libjulia/optimiser/FunctionHoister.cpp:43]: Might be just a typo, but does it make a difference?

Comment (pirapira) [reply libjulia/optimiser/FunctionHoister.cpp:43]: I thought the C++ compilers should fail, but they don't, so I don't know.

Comment (chriseth) [reply libjulia/optimiser/FunctionHoister.cpp:43]: `auto&&` has a special meaning: https://stackoverflow.com/questions/13230480/what-does-auto-tell-us

Comment (pirapira) [reply libjulia/optimiser/FunctionHoister.cpp:43]: That's weird.  I'll check the spec (well, something close to it) and file a complaint.

File: libjulia/optimiser/FunctionHoister.h (hunk 2 above)
Comment (chriseth) [reply libjulia/optimiser/FunctionHoister.h]: Added.

**Review by axic — APPROVED** [review]

**Pull request merged**